File: Ch03/teste.hs
class Monoid' m where
    mempty'  :: m
    mappend' :: m -> m -> m
    
instance Monoid' Int where
    mempty' = 1
    mappend' = (*)
    
main = do
  let x = 2 :: Int
      y = 3 :: Int
  print (mappend' x y)

File: Ch03/teste.cpp
#include <iostream>
#include <string>
#include <cstddef>

template<class T>
  T mempty;

template<class T>
  T mappend(T, T);

template<class M>
  concept bool Monoid = requires (M m) {
    { mempty<M> } -> M;
    { mappend(m, m) } -> M;
  };
  
template<>
int mempty<int> = {1};

int mappend(int x, int y) {
    return x*y;
}

int main()
{
  std::cout << mappend(2,3);
}
